### PR TITLE
feat: return precalculated address from deploy, reimplement multi-deploy

### PR DESCRIPTION
**Pull request by tabaktoni**
## Motivation and Resolution
Fix https://github.com/0xs34n/starknet.js/issues/442
add the precalculated address on the deploy return.

## Usage-related changes
added support for multi-deploy with singular payload. We had it before but the user would need to open the source code copy function body and basically create a call by himself and pass that to the additional parameter.

## Checklist:

- [x] Performed a self-review of the code
- [x] Rebased to the last commit of the target branch (or merged it into my branch)
- [x] Linked the issues which this PR resolves
- [x] Documented the changes
- [x] Updated the docs (www)
- [x] Updated the tests
- [x] All tests are passing

### CODE REVIEW

File: __tests__/account.test.ts [hunk 1]
```diff
@@ -289,12 +290,55 @@ describe('deploy and test Wallet', () => {
...
       expect(deployment).toHaveProperty('transaction_hash');
+
+      // check pre-calculated address
+      const txReceipt = await provider.waitForTransaction(deployment.transaction_hash);
+      const udcEvent = parseUDCEvent(txReceipt);
+      expect(cleanHex(deployment.contract_address[0])).toBe(cleanHex(udcEvent.contract_address));
+    });
+
+    test('UDC multi Deploy', async () => {
+      // TODO: add test with multiple deploys
```
Comment (ivpavici) [inline __tests__/account.test.ts:325]: is this todo leftover or?

Comment (tabaktoni) [reply __tests__/account.test.ts:325]: Yep :D

Comment (tabaktoni) [inline __tests__/account.test.ts:325]:
```suggestion
```

File: src/account/default.ts [hunk 2]
```diff
@@ -321,22 +321,22 @@ export class Account extends Provider implements AccountInterface {
...
-  ): Promise<InvokeFunctionResponse> {
-    const compiledConstructorCallData = compileCalldata(constructorCalldata);
-    const callsArray = Array.isArray(additionalCalls) ? additionalCalls : [additionalCalls];
-    const deploySalt = salt ?? randomAddress();
-
-    return this.execute(
-      [
-        {
+    payload: UniversalDeployerContractPayload | UniversalDeployerContractPayload[],
+    details?: InvocationsDetails | undefined
```
Comment (penovicp) [inline src/account/default.ts:325]: Nitpick, `| undefined` should be covered by the parameter being optional. Noticed it for `estimateDeployFee`, `deployContract`, and `declareDeploy` as well.

**Review by penovicp — APPROVED** [review]

File: src/account/default.ts (hunk 2 above)
Comment (tabaktoni) [reply src/account/default.ts:325]: Yea but I had some issue with that so I defined it as also undefined so TS stop crying.

**Pull request merged**